lib/build-components/build_components/templates/python/macros.j2

```diff
@@ -0,0 +1,180 @@
+{# [Builder image]
+    Download and install poetry under the root user.
+#}
+{% macro install_poetry(version) %}
+# Install poetry
+# Note: We prefer to install python packages as --user so that the artifacts are easier
+#       to copy out of the builder image into the final image. poetry already installs
+#       itself under the user directory, and we eventually copy the poetry tool and the
+#       installed python packages under the image USER in the resulting image.
+
+# Ensure we have curl so we can install poetry
+RUN apt-get update \
+ && apt-get -y --no-install-recommends install \
+    curl
+
+# Configure pip, mainly to ensure that packages are installed in the user directory.
+COPY pip.conf /etc/pip.conf
+
+# Install poetry under the root user's home directory.
+# On some images "sh" is aliased to "dash" which does not support "set -o pipefail".
+# We use the "exec" form of RUN to delegate this command to bash instead.
+# This is all because we have a pipe in this command and we wish to fail the build
+# if the any command in the pipeline fails.
+RUN ["/bin/bash", "-c", "set -o pipefail && pip install --upgrade pip && curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | POETRY_VERSION={{version}} python && chmod a+x /root/.poetry/bin/poetry"]
```

Comment:
Why don't we want to follow this same pattern for the top-level Dockerfile?  And why is it necessary to explicitly set the poetry executable executable?  Does get-poetry.py not do that?

Reply:
> Why don't we want to follow this same pattern for the top-level Dockerfile?

We probably should.

> And why is it necessary to explicitly set the poetry executable executable? Does get-poetry.py not do that?

The important thing wasn't setting it executable, it was setting it executable for everyone. Fortunately, I've figured out a way to obviate the need for this.